[EDNA-89] Support sorting of experiments by more fields

gromakovsky opened the pull request:
## Description

Problem: it's possible to sort experiments by upload date, but there
are other columns in the table with experiments and it often makes
sense to sort by them as well.

Solution:
- [x] Support sorting by compound, target and methodology names.
- [x] Add these names to `ExperimentResp`.
- [x] Add implicit base sorting by IDs, so that items with equal sorting keys (or all items when there is no sorting) always have deterministic order.
- [x] Update frontend according to new (extended) `ExperimentResp` structure.

## Related issue(s)

<!--
- Short description of how the PR relates to the issue, including an issue link.
For example
- Fixed #100500 by adding lenses to exported items

Write 'None' if there are no related issues (which is discouraged).
-->

https://issues.serokell.io/issue/EDNA-89

## :white_check_mark: Checklist for your Pull Request

<!--
Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

If you don't set a checkmark (e. g. don't add a test for new functionality),
you must be able to justify that.
-->

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

gromakovsky committed:
Support sorting of experiments by methodology name

Problem: it's possible to sort experiments by upload date, but there
are other columns in the table with experiments and it often makes
sense to sort by them as well.

Solution:
1. Add "methodology" to `SortingParamTypesOf ExperimentResp`, thus
making it possible to pass `+methodology` sorting order for example.
2. Update `getExperiments` query to get the whole `TestMethodology`
record and include it into `sortingApp`.
3. Add default sorting by ID to make order of entries with equal
methodology name defined precisly.
4. Methodology name is now included into `ExperimentResp` along with
methodology ID.

gromakovsky committed:
Support sorting of experiments by compound and target

Problem: we've added sorting of experiments by test methodology,
but there are other columns for which sorting can be useful and
is not supported yet.

Solution: add sorting by compound and target similarly to methodology.
Sorting by IC50 can be supported too, but with some extra tricks.
So it's postponed for future work, let's see whether it will be
requested.

gromakovsky committed:
Update frontend to handle new ExperimentResp

Problem: ExperimentResp in edna-server was updated and now contains
names of compound, target and methodology.
Frontend still expects only IDs.

Solution: update `ExperimentDto` to contain new data, use this new data
in `filteredExperimentsQuery`.

gromakovsky committed:
Add default sorting order to Library queries

Problem: if you use Library getters returning lists without sorting,
you may get somewhat non-deterministic results because PostgreSQL
doesn't guarantee any specific order if you don't request it
explicitly.
Non-deterministic results are inconvenient to deal with, especially
with pagination.

Solution: use IDs as fallback sorting keys (when all other sorting
keys are equal). We are using `sortingSpecWithId` like we did for
Dashboard.

gromakovsky committed:
Support sorting by methodology, compound and target on frontend

Problem: backend now supports sorting by methodology, compound and
target, but this functionality is not used by frontend yet.

Solution: update relevant columns in frontend code by remove
`disableSortBy` and defining proper `id`s.

pva701 commented on frontend/src/api/types.ts:
```diff
@@ -94,9 +94,9 @@ export type CompoundDto = {
-  compound: number;
-  target: number;
-  methodology?: number;
+  compound: [number, string];
```
why do you need `id` here?

gromakovsky replied:
I am not sure if we need it, but I decided to keep it here because it's very small and may be useful in future. This way the response is more self-contained.

sashasashasasha151 committed:
Fix dashboard table renderring

Problem: after sorting addition to the dashboard table the arrow
we show for sorting is shown below column name and it doesn't look
nice, also doesn't match design.

Solution: make some css hacks to make it look properly. The problem
was caused by the fact that sometimes IC50 values take too much width,
e. g. if the value is grater than 1e6 or something like that.
We force the value to never exceed a certain width by trimming it and
inserting ellipsis when it exceeds the limit.

worm2fed commented on backend/src/Edna/Dashboard/DB/Query.hs:
```diff
@@ -135,19 +164,29 @@ getExperiments mProj mComp mTarget sorting pagination =
   where
-    sortingApp ((_experimentRec, _projectId, _methodologyId, uploadDate, _subExpId), _) =
+    eExperimentId' :: ExperimentRec -> SqlSerial Word32
+    eExperimentId' = eExperimentId
```
Why do we need it?

gromakovsky replied:
I was getting a weird compilation error involving complex Beam types without it :shrug: 

Let me check whether it's still necessary.

gromakovsky replied:
No longer necessary, removed.

worm2fed commented on backend/src/Edna/Orphans.hs:
```diff
@@ -28,3 +29,9 @@ instance ApiHasArgClass (MultipartForm' mods tag t) where
+
+-- defined the same way as in @Lens.Micro.Internal@ (where instances are
+-- provided only for tuples with up to 5 items)
+instance Field1 (a,b,c,d,e,f,g) (a',b,c,d,e,f,g) a a' where
```
```suggestion
instance Field1 (a, b, c, d, e, f, g) (a', b, c, d, e, f, g) a a' where
```

worm2fed commented on backend/src/Edna/Orphans.hs:
```diff
@@ -28,3 +29,9 @@ instance ApiHasArgClass (MultipartForm' mods tag t) where
+-- defined the same way as in @Lens.Micro.Internal@ (where instances are
+-- provided only for tuples with up to 5 items)
+instance Field1 (a,b,c,d,e,f,g) (a',b,c,d,e,f,g) a a' where
+  _1 k ~(a,b,c,d,e,f,g) = (,b,c,d,e,f,g) <$> k a
```
```suggestion
  _1 k ~(a, b, c, d, e, f, g) = (, b, c, d, e, f, g) <$> k a
```

gromakovsky committed:
Improve formatIC50 function

Problem: after the last change we often show ellipsis for IC50
values to keep the IC50 column not too big. Values with ellipsis
look ugly and even inconsistently. Even similar values like
4.3661 and 5.7405 can be shown differently (with ellipsis and without).

Solution: update `formatIC50` trying to never produce a string with
more than 5 characters. We empirically noticed that 5 characters
always fit into the column without ellipsis. We still keep the limiting
logic just in case `formatIC50` produced a too long string, but we
hope it will never be triggered.

worm2fed commented on frontend/src/utils/utils.ts:
```diff
@@ -49,8 +49,49 @@ export function zip<A, B>(a: A[], b: B[]): [A, B][] {
+  // We are carefully listing all cases here to think about each of them.
+  // Perhaps it can be optimized, e. g. using 'toPrecision', but I (@gromak)
+  // decided to follow the KISS principle.
+  if (x >= 1e5) {
```
Maybe it's better to use switch or lookup map?
https://betterprogramming.pub/stop-putting-so-many-if-statements-in-your-javascript-3b65aaa4b86b

gromakovsky replied:
How to use `switch` here? AFAIU switch works when you compare with some values for equality, but here I compare only for `>=`. Same for lookup map. I am new to JS/TS, so maybe I'm missing something, but it's not clear to me how to use `switch` here.

worm2fed replied:
switch:
```
values=[1000,  2000 ... 29000, 30000];

switch(true) {
  case (val < values[0]): /* do something */ break;
  case (val < values[1]): /* do something */ break;
  ...
  case (val < values[29]): /* do something */ break;
}
```

with lookup map.. maybe there are option to use expressions as keys.. but maybe I'm wrong and this is from other language

gromakovsky replied:
Hm, ok, I'll give it a try.

worm2fed replied:
it's good now (and faster)
